### PR TITLE
Added hide and error message when a user tries to create a challenge for HabitRPG group and they have no gems

### PR DESCRIPTION
**Pull request by TheHollidayInn**
This is a fix for https://github.com/HabitRPG/habitrpg/pull/4836 . 
-- I have move the scope of the enough gems variable up.
-- I have added a check to make sure there is a challenge being created before displaying an error message

### CODE REVIEW

**Comment by Alys** [comment]
Editing works well now!

If you have zero gems and try to create a Tavern challenge, the message "You do not have enough gems to post this challenge" appears, as desired. However if you change your mind and try to create it in a different guild, the message stays there. Is it possible to make the message disappear when the selected guild is changed?

**Comment by TheHollidayInn** [comment]
Umm... I can definitely make that change. I seem to have messed up with rebasing..

**Comment by TheHollidayInn** [comment]
Ah, ha! I fixed it.

**Comment by TheHollidayInn** [comment]
@crookedneighbor I'm having trouble writing tests for this. I have to test a $watch, and after reviewing the documentation, it seems I need to call $digest or $apply, but both are causing this error in testing:

```
Unexpected request: GET /api/v2/challenges
    No more request expected
    Error: Unexpected request: GET /api/v2/challenges
    No more request expected
```

**Comment by crookedneighbor** [comment]
@TheHollidayInn Post the test you have and I'll troubleshoot it with you.

**Comment by TheHollidayInn** [comment]
```
it('should prevent user from creating Tavern challenge if they have 0 gems', function() {
   scope.create();
   scope.$apply();
   scope.save(scope.obj);
  });
```

**Comment by crookedneighbor** [comment]
Realized after you sent me your test that there was no existing test file for the Challenges controller, so I made one: https://github.com/HabitRPG/habitrpg/blob/blade-TheHollidayInn-fix-max-value-0-challenges/test/spec/challengeCtrlSpec.js

You can pull it in with this:

``` bash
git pull upstream blade-TheHollidayInn-fix-max-value-0-challenges
```

They aren't the best unit tests in the world, but it's a start. Writing it made me realize what a mess the save function in the Challenges Ctrl is.

**Comment by TheHollidayInn** [comment]
Yea, I created the new challengeCtrlSpec file in the PR about sorting challenges by owner. I have started adding other challenge test and noticed some weird things as well. I'll add in your tests. Thanks for the help!

**Comment by TheHollidayInn** [comment]
Hey, @crookedneighbor , I don't see your Unit tests anymore.

**Comment by crookedneighbor** [comment]
I thought you were done with the branch, so I deleted it. I've restored it now.

**Comment by TheHollidayInn** [comment]
The tests are great. However, I am trying to test the `$scope.$watch('newChallenge.group'` function, but I can't seem to get the watch to trigger in the tests. I want to make sure that when a user selects habitrpg group the enoughGems is set to the correct value.

```
describe("enough gems tracking", function() {

    it("defaults enoughGems to true", function() {
      expect(scope.enoughGems).to.equal(true);
    });

    it("set enoughGems to false when user selects habitrpg without enough gems", function() {
      scope.newChallenge = new Challenges.Challenge({
        name: 'Challenge without a group',
        description: '',
        habits: [],
        dailys: [],
        todos: [],
        rewards: [],
        leader: user._id,
        timestamp: +(new Date),
        members: [],
        official: false
      });
      scope.newChallenge.group = 'habitrpg';
      scope.$digest()
      expect(scope.enoughGems).to.equal(false);
    });

    it("set enoughGems to true when user selects habitrpg with enough gems", function() {
      scope.newChallenge = new Challenges.Challenge({
        name: 'Challenge without a group',
        description: '',
        habits: [],
        dailys: [],
        todos: [],
        rewards: [],
        leader: user._id,
        timestamp: +(new Date),
        members: [],
        official: false
      });
      user.balance = 1;
      scope.newChallenge.group = 'habitrpg';
      scope.$digest()
      expect(scope.enoughGems).to.equal(true);
    });

  });
```

**Comment by TheHollidayInn** [comment]
Alright, how about this. I removed the dependency on checking the watch, and I added tests that are more dependent on the user's balance to verify the enougGems variable gets used correctly.

File: test/spec/challengeCtrlSpec.js [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        todos: [],
+        rewards: [],
+        leader: user._id,
+        timestamp: +(new Date),
+        members: [],
+        official: false
+      });
+      user.balance = 1;
+      scope.newChallenge.group = 'habitrpg';
+      //expect(scope.enoughGems).to.equal(true);
```
Comment (crookedneighbor) [inline test/spec/challengeCtrlSpec.js]: Is this intended to be commented out?

**Comment by TheHollidayInn** [comment]
There we go. I changed my test to use the new hasEnoughGems function.

File: website/public/js/controllers/challengesCtrl.js [hunk 2]
```diff
@@ -277,6 +279,28 @@ habitrpg.controller("ChallengesCtrl", ['$rootScope','$scope', 'Shared', 'User',
       $scope.search.group = _.transform($scope.groups, function(m,g){m[g._id] = false});
     }
 
+    $scope.shouldShow = function(task, list, prefs){
+      return true;
+    };
+
+    //A helper function to dermine if the user has enough gems to create a Habitrpg challenge
+    $scope.hasEnoughGems = function(gid) {
```
Comment (crookedneighbor) [inline website/public/js/controllers/challengesCtrl.js]: If this is a helper function, would it make more sense to make it a private method?

File: website/views/options/social/challenges.jade [hunk 3]
```diff
@@ -128,12 +128,25 @@ script(type='text/ng-template', id='partials/options.social.challenges.html')
...
                 textarea.form-control(cols='3', placeholder=env.t('challengeDescr'), ng-model='newChallenge.description')
 
+<<<<<<< HEAD
               .form-group
                 .input-group
                   span.input-group-addon
                     .Pet_Currency_Gem1x
                   input.form-control(type='number', min="{{newChallenge.group=='habitrpg' ? 1 : 0}}", max="{{maxPrize}}", ng-model='newChallenge.prize', placeholder=env.t('prize'))
                 a.hint(popover="{{newChallenge.group=='habitrpg' ? env.t('prizePopTavern') : env.t('prizePop')}}", popover-trigger='mouseenter', popover-placement='right')
+=======
```
Comment (crookedneighbor) [inline website/views/options/social/challenges.jade]: Looks like you forgot to sort out a merge conflict.

File: website/public/js/controllers/challengesCtrl.js (hunk 2 above)
Comment (TheHollidayInn) [inline website/public/js/controllers/challengesCtrl.js]: That makes sense. I'm unsure of which is better. I could see use for this on both sides. Is there a benefit to taking the function off scope?

File: website/views/options/social/challenges.jade (hunk 3 above)
Comment (TheHollidayInn) [inline website/views/options/social/challenges.jade]: Hmm. That's weird. I'm not sure how this got passed when I use --continue, but I was having some trouble. I'll push this fix after we decide on the above function.

**Comment by TheHollidayInn** [comment]
@crookedneighbor I just realized I could rebase without making a new commit.

The state transition doesn't seemed to be being called. Any pointers on how I can get that to work?

**Comment by TheHollidayInn** [comment]
Alright, I updated the test to resemble the other save test which changes a variable when the mock is called. I think the test is broken because the challenge._locked depends on a change in the state url.

File: website/public/js/controllers/challengesCtrl.js [hunk 4]
```diff
@@ -314,6 +316,28 @@ habitrpg.controller("ChallengesCtrl", ['$rootScope','$scope', 'Shared', 'User',
       $scope.search.group = _.transform($scope.groups, function(m,g){m[g._id] = false});
     }
 
+    $scope.shouldShow = function(task, list, prefs){
+      return true;
```
Comment (crookedneighbor) [inline website/public/js/controllers/challengesCtrl.js]: Wait.... what does this do?

Comment (crookedneighbor) [inline website/public/js/controllers/challengesCtrl.js]: Ahh, I see. You accidentally duplicated this from below, and the function below overwrites the shouldShow function from the tasks controller in the view when used with the challenges controller.

**Comment by TheHollidayInn** [comment]
Sweet. Updated with a new name and removed the extra function.

**Comment by crookedneighbor** [comment]
I rebased and merged this in after doing some work on it. Notated toward your next tier.

**Comment by TheHollidayInn** [comment]
Sweet. Thanks!

**Pull request closed without merging**